### DISCUSSION TITLE
v4.0.0-M2

### DISCUSSION THREAD

**Discussion by graemerocher** [micronaut-projects/micronaut-core#9106] [General]
### New Features 🎉
* Add support for annotation-based CORS configuration (8558) by @wetted in https://github.com/micronaut-projects/micronaut-core/pull/8580
* Support for Compilation Time Expressions in Annotations by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/8954
* add ability to disable streaming http request processing by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9006
* Support for conditional routing by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9094
### Bug Fixes 🐞
* Logback configuration defined via JAVA_TOOL_OPTIONS is ignored by @timyates in https://github.com/micronaut-projects/micronaut-core/pull/9009
### Improvements ⭐
* tck: static resource test by @sdelamo in https://github.com/micronaut-projects/micronaut-core/pull/8971
* Unwrap optionals with safe navigation in expressions by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9024
* Use `Map.of` instead of custom implementation by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/9032
### Docs 📖
* doc: don't use Micronaut as a noun by @sdelamo in https://github.com/micronaut-projects/micronaut-core/pull/8958
* Fix broken link to checkstyle issue img by @Januson in https://github.com/micronaut-projects/micronaut-core/pull/9047
### Dependency updates 🚀
* Bump Jib Maven Plugin version by @alvarosanchez in https://github.com/micronaut-projects/micronaut-core/pull/8980
* build: netty 4.1.90-Final by @sdelamo in https://github.com/micronaut-projects/micronaut-core/pull/9019
* fix(deps): update managed-jackson to v2.14.2 by @renovate in https://github.com/micronaut-projects/micronaut-core/pull/8169
* fix(deps): update managed-slf4j to v2.0.7 by @renovate in https://github.com/micronaut-projects/micronaut-core/pull/8537
* chore(deps): update plugin com.google.devtools.ksp to v1.8.20-1.0.10 by @renovate in https://github.com/micronaut-projects/micronaut-core/pull/8676
* fix(deps): update dependency org.testcontainers:spock to v1.17.6 by @renovate in https://github.com/micronaut-projects/micronaut-core/pull/9044
* build: Kotlin 1.8.20 by @sdelamo in https://github.com/micronaut-projects/micronaut-core/pull/9055
* chore(deps): update gradle/gradle-build-action action to v2.4.0 by @renovate in https://github.com/micronaut-projects/micronaut-core/pull/9080
* fix(deps): update netty monorepo to v4.1.91.final by @renovate in https://github.com/micronaut-projects/micronaut-core/pull/9079
* fix(deps): update junit5 monorepo by @renovate in https://github.com/micronaut-projects/micronaut-core/pull/9078
### Other Changes 💡
* Fix bug preventing scheduled jobs from being ran in the specified time zone by @Gradsta in https://github.com/micronaut-projects/micronaut-core/pull/8911
* More mTLS test cases by @yawkat in https://github.com/micronaut-projects/micronaut-core/pull/8944
* Move from `javax.el` to `jakarta.el`. by @altro3 in https://github.com/micronaut-projects/micronaut-core/pull/8966
* build: Apache Groovy 4.0.10 by @sdelamo in https://github.com/micronaut-projects/micronaut-core/pull/8975
* Embedded HTTP benchmark, various small perf improvements by @yawkat in https://github.com/micronaut-projects/micronaut-core/pull/8974
* Fix validation tests by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/8983
* Correct KSP parameters matching by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/8989
* Correct KotlinClassElement's `isAssignable` by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/8988
* add missing logback.xml by @wetted in https://github.com/micronaut-projects/micronaut-core/pull/8987
* Fix race condition in FilterRunner by @yawkat in https://github.com/micronaut-projects/micronaut-core/pull/9005
* Make Kotlin and Kotlin coroutines versions managed. by @wetted in https://github.com/micronaut-projects/micronaut-core/pull/8995
* More performance improvements by @yawkat in https://github.com/micronaut-projects/micronaut-core/pull/8993
* fix java.lang.IllegalArgumentException: duplicate element: ALL_PUBLIC_METHODS by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9015
* Correct typo in `OutgointRequestProcessorMatcher` classname. by @wetted in https://github.com/micronaut-projects/micronaut-core/pull/9016
* Support JDK 20 in annotation processors by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/9022
* Improve support for logback 1.3+ by @timyates in https://github.com/micronaut-projects/micronaut-core/pull/9018
* Enhance TCK server headers test by @timyates in https://github.com/micronaut-projects/micronaut-core/pull/9025
* Cleanup expressions implementation and improve perf by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9027
* ci: fold release-notes & sonar in gradle workflow by @sdelamo in https://github.com/micronaut-projects/micronaut-core/pull/9033
* Make evaluation context objects effectively singleton by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9034
* Add a class for handling headers in a case-insensitive way by @timyates in https://github.com/micronaut-projects/micronaut-core/pull/9031
* Update CRaC to 1.2.1 for Micronaut 3.9.0 by @timyates in https://github.com/micronaut-projects/micronaut-core/pull/8992
* Fix broken incremental compilation with Gradle by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9038
* Retrieving beans from bean context in expressions by @GavrilovSV in https://github.com/micronaut-projects/micronaut-core/pull/9041
* Merge branch 3.9.x into 4.0.x by @sdelamo in https://github.com/micronaut-projects/micronaut-core/pull/9042
* Bump Spock to 2.3 and Groovy to 4.0.11 by @szpak in https://github.com/micronaut-projects/micronaut-core/pull/9054
* Retrieving environment properties in expressions by @GavrilovSV in https://github.com/micronaut-projects/micronaut-core/pull/9053
* More performance optimizations by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/9070
* Performance optimizations: cache executor, optimize HttpMethod enum by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9063
* Fix Groovy/Spock bug in Kubernetes module by @graemerocher in https://github.com/micronaut-projects/micronaut-core/pull/9067
* Microoptimizations by @yawkat in https://github.com/micronaut-projects/micronaut-core/pull/9069
* Make legacy filters work like in Micronaut 3 by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/9071
* KSP: Ignore class element nullability for isAssignable by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/9075
* Expose ConnectionManager for use in Oracle Cloud Netty client by @yawkat in https://github.com/micronaut-projects/micronaut-core/pull/9073
* Issue 8592 - Update Custom converter docs to use ConversionService bean by @wetted in https://github.com/micronaut-projects/micronaut-core/pull/8638
* Disable leak detection by default by @yawkat in https://github.com/micronaut-projects/micronaut-core/pull/9074
* More performance optimizations by @dstepanov in https://github.com/micronaut-projects/micronaut-core/pull/9076
* Fix buffer leak by @yawkat in https://github.com/micronaut-projects/micronaut-core/pull/9092
* use Milestone versions by @sdelamo in https://github.com/micronaut-projects/micronaut-core/pull/9085
* Remove unused logback.xml file by @timyates in https://github.com/micronaut-projects/micronaut-core/pull/9104

## New Contributors
* @Gradsta made their first contribution in https://github.com/micronaut-projects/micronaut-core/pull/8911
* @Januson made their first contribution in https://github.com/micronaut-projects/micronaut-core/pull/9047

**Full Changelog**: https://github.com/micronaut-projects/micronaut-core/compare/v4.0.0-M1...v4.0.0-M2

<hr /><em>This discussion was created from the release <a href='https://github.com/micronaut-projects/micronaut-core/releases/tag/v4.0.0-M2'>v4.0.0-M2</a>.</em>

**Comment by subash89** [comment]
@graemerocher   Micronaut-parent 4.0.0-M2 is not available in public maven repo, only 4.0.0-M1 is there. All other micronaut poms and jars available in 4.0.0-M2. Assume it's missed when deploy ?

**Reply by graemerocher** [reply]
The parent Pom moved to the platform project under the group io.micronaut.platform

**Reply by subash89** [reply]
Thanks for the update. platform pom M1 worked. Is there any estimated timelines for the 4.0.0 GA Release date

**Reply by graemerocher** [reply]
We are trying for this quarter

**Reply by alvarosanchez** [reply]
> The parent Pom moved to the platform project under the group io.micronaut.platform

That's not the case. What happens here is that what was known as "the Micronaut version" is now [Micronaut Platform](https://github.com/micronaut-projects/micronaut-platform), whose latest version is 4.0.0-M2.

This release (4.0.0-M1) is for Micronaut Core, which is now versioned independently of the Micronaut Platform version.

Regarding the Parent POM, for historical reasons it is still under `io.micronaut:micronaut-parent`, but I'm gonna change it to `io.micronaut.platform:micronaut-parent`, since it is versioned along with Micronaut Platform.

**Reply by subash89** [reply]
@alvarosanchez  But micronaut-core is 4.0.0-M2, micronaut-platform is 4.0.0-M1 latest versions as of today

**Reply by alvarosanchez** [reply]
Correct

**Reply by subash89** [reply]
Just FYI since we on the 4.0.0-M2 topic, having the parent pom io.micronaut.platform:micronaut-platform:4.0.0-M1 defined in my maven project, with property exec.mainClass, the jar it creates does not have main class defined in the manifest. So you can't really start the app with java -jar with that. This was not the case in 3.x version.

**Reply by alvarosanchez** [reply]
`io.micronaut.platform:micronaut-platform:4.0.0-M1` is the _Platform BOM_. The Parent POM for Micronat 4.x will be at `io.micronaut.platform:micronaut-parent`

**Reply by subash89** [reply]
Is it locally and not yet deployed to maven repos? 
only micronaut-platform is there under group id io.micronaut.platform
https://mvnrepository.com/search?q=io.micronaut.platform

**Reply by subash89** [reply]
With parent pom 4.0.0-M1 I cannot build the project, following error.  It's looking for micronaut-platform 4.0.0-SNAPSHOT. 
**I found the reason:**
 https://github.com/micronaut-projects/micronaut-maven-plugin/blob/v4.0.0-M1/pom.xml pointing to wrong micronaut version. 
This means micronaut-maven-plugin release branch cannot be built on a clean m2 repo. I believe you able to build this some how the SNAPSHOT is published to your local remote repos.

  Unresolveable build extension: Plugin io.micronaut.maven:micronaut-maven-plugin:4.0.0-M1 or one of its dependencies could not be resolved: Failed to read artifact descriptor for io.micronaut.maven:micronaut-maven-plugin:jar:4.0.0-M1: Could not find artifact io.micronaut.platform:micronaut-platform:pom:4.0.0-SNAPSHOT ->

**Reply by subash89** [reply]
Created a PR with the change https://github.com/micronaut-projects/micronaut-maven-plugin/pull/710

**Reply by subash89** [reply]
I saw a supersedded  fix is merged. Is there nightly builds or snapshot version we can use ? 
Because local build did not work.  I tried to only build maven plugin project and use 4.0.0-SNAPSHOT version to see the local changes.
I am using 4.0.0-M1 micronaut-parent pom.
 
Fatal error compiling: java.lang.NoSuchMethodError: Micronaut method io.micronaut.context.AbstractInitializableBeanDefinition.getEvaluatedExpressionValueForMethodArgument(int,int) not found. Most likely reason for this issue is that you are running a newer version of Micronaut with code compiled against an older version. Please recompile the offendin

**Reply by alvarosanchez** [reply]
Right, so now there are a few moving parts that we need to settle and stabilise.

There are Snapshot versions published, of course, see https://docs.micronaut.io/latest/guide/#usingsnapshots 

I recommend you to check any of the examples at https://github.com/micronaut-projects/micronaut-maven-plugin/tree/master/examples

**Reply by subash89** [reply]
I did, it's pretty much same. But SNAPSHOT build is tricky, I get missing snapshot errors for some when build. Which means I have to build all related micronaut projects locally. 

**Are you planning to release  next milestone release any sooner ?**

**Reply by graemerocher** [reply]
we are working on the CLI/starter which is the last piece for everything to be released together. Things will start stabilising in the next couple of weeks

**Reply by alvarosanchez** [reply]
You don't need local builds of anything.
